File: android/src/main/java/com/readapp/data/ApiBackend.kt
package com.readapp.data

enum class ApiBackend {
    Read,
    Reader
}

fun detectApiBackend(serverUrl: String): ApiBackend {
    val normalized = serverUrl.lowercase()
    return when {
        normalized.contains("/reader3") -> ApiBackend.Reader
        normalized.contains("/api/") -> ApiBackend.Read
        else -> ApiBackend.Read
    }
}

fun normalizeApiBaseUrl(serverUrl: String, backend: ApiBackend): String {
    if (serverUrl.isBlank()) return serverUrl
    val trimmed = serverUrl.trimEnd('/')
    return when (backend) {
        ApiBackend.Reader -> if (trimmed.contains("/reader3")) trimmed else "$trimmed/reader3"
        ApiBackend.Read -> if (trimmed.contains("/api/")) trimmed else "$trimmed/api/5"
    }
}

fun stripApiBasePath(serverUrl: String): String {
    val trimmed = serverUrl.trimEnd('/')
    return trimmed
        .replace(Regex("/api/\\d+$"), "")
        .replace(Regex("/reader3$"), "")
        .trimEnd('/')
}

File: ios/ReadApp/Services/ApiBackend.swift
import Foundation

enum ApiBackend: String, CaseIterable, Identifiable {
    case read
    case reader

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .read:
            return "轻阅读"
        case .reader:
            return "阅读3"
        }
    }
}

struct ApiBackendResolver {
    static func detect(from serverURL: String) -> ApiBackend {
        let normalized = serverURL.lowercased()
        if normalized.contains("/reader3") {
            return .reader
        }
        if normalized.contains("/api/") {
            return .read
        }
        return .read
    }

    static func normalizeBaseURL(_ serverURL: String, backend: ApiBackend, apiVersion: Int) -> String {
        var trimmed = serverURL.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }
        switch backend {
        case .reader:
            if trimmed.contains("/reader3") {
                return trimmed
            }
            return "\(trimmed)/reader3"
        case .read:
            if trimmed.contains("/api/") {
                return trimmed
            }
            return "\(trimmed)/api/\(apiVersion)"
        }
    }

    static func stripApiBasePath(_ baseURL: String) -> String {
        var trimmed = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }
        if let range = trimmed.range(of: "/api/\\d+$", options: .regularExpression) {
            trimmed.removeSubrange(range)
        } else if trimmed.hasSuffix("/reader3") {
            trimmed.removeLast("/reader3".count)
        }
        return trimmed
    }
}
